pkg/registry/apis/provisioning/jobs/changes.go

```diff
@@ -0,0 +1,89 @@
+package jobs
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+
+	provisioning "github.com/grafana/grafana/pkg/apis/provisioning/v0alpha1"
+	"github.com/grafana/grafana/pkg/registry/apis/provisioning/repository"
+	"github.com/grafana/grafana/pkg/registry/apis/provisioning/resources"
+)
+
+type ResourceFileChange struct {
+	// Path to the file in a repository with a change
+	Path   string
+	Action repository.FileAction
+
+	// The current value in the database -- only required for delete
+	Existing *provisioning.ResourceListItem
+}
+
+func Changes(source []repository.FileTreeEntry, target *provisioning.ResourceList) ([]ResourceFileChange, error) {
+	lookup := make(map[string]*provisioning.ResourceListItem, len(target.Items))
+	for _, item := range target.Items {
+		if item.Path == "" {
+			if item.Group != "folder.grafana.app" {
+				return nil, fmt.Errorf("empty path on a non folder")
+			}
+			continue
+		}
+		lookup[item.Path] = &item
+	}
+
+	var keep []string
+	changes := make([]ResourceFileChange, 0, len(source))
+	for _, file := range source {
+		if !file.Blob {
+			continue // skip folder references?
+		}
+
+		check, ok := lookup[file.Path]
+		if ok {
+			if check.Hash != file.Hash {
+				changes = append(changes, ResourceFileChange{
+					Action:   repository.FileActionUpdated,
+					Path:     check.Path,
+					Existing: check,
+				})
+			}
+			keep = append(keep, file.Path)
+			delete(lookup, file.Path)
+		} else if !resources.ShouldIgnorePath(file.Path) {
+			changes = append(changes, ResourceFileChange{
+				Action: repository.FileActionCreated, // or previously ignored/failed
+				Path:   file.Path,
+			})
+		}
+	}
+
+	// nested nested loop :grimmice: trie?
```

Comment:
this precisely what the resource tree was doing with the `tree.In`. The tree was built out of the resources and why I wanted to the tree to work only for its tree properties and not be about k8s / clients.

Reply:
dug deeper here... tree.In does not help find matching *prefix* -- so i will keep as is for now

I am pretty sure we will need a tree structure that has real parent/child relationships -- as implemented the `FolderTree` is a lookup function, but does really know about the structure